[14.0][MIG] l10n_es_aeat_mod349: migration to 14.0

miguel-S73 opened the pull request:
Standard migration to 14.0
&&
Cherry-pick [#1579](https://github.com/OCA/l10n-spain/pull/1579)

Commit:
[ADD] l10n_es_aeat, l10n_es_aeat_347, l10n_es_aeat_349: Adds a generic module for aeat models, 347 module was portedto v6.0 and adds new module to print AEAT model 349.

Commit:
[IMP+FIX] l10n_es_aeat, l10n_es_aeat_mod347, l10n_es_aeat_mod349: Pequeñas mejoras generales (vistas, traducciones, código) y corrección pequeño bug por un olvido en la adaptación de la v5 a la v6

albertca committed:
[FIX] l10n_es_aeat_mod349: 3 things:

* Only use the first four digits of fiscal year's code to determine fiscal year's number and also give a decent message to the user if it's not a number instead of simply crashing.
* Hacer opcional el campo Operation Key en las facturas, permitiendo que esté en blanco.
* Ser un poco menos rigurosos con que el año esté en el código del año fiscal. Atención: deberíamos añadir un campo con este valor en lugar de hacer este hack.

Commit:
[FIX+RFR] l10n_es_aeat_mod349: 2 things:

* FIX: Corregimos una serie de errores que había en las rectificaciones a la hora de obtener las declaraciones originales, si había alguna que cumplía los requisitos pero sin pertenecer a un modelo. Quizás se solucionara mejor con un on_delete='cascade'.
* RFR: Cambio de licencia de GPL3 a AGPL3 en el resto de módulos de localización española, con el consentimiento de todos los autores expresado vía lista de correo del equipo de localización openerp-spain@googlegroups.com

albertca committed:
[FIX] l10n_es_aeat_mod349: Several fixes

* Only use the first four digits of fiscal year's code to determine fiscal year's number and also give a decent message to the user if it's not a number instead of simply crashing.
* Resolve a problem with python isinstance which considers 'False' is an integer and which makes OpenERP create operation_key field as integer instead of char.
* Convert a return None to return False, because the XML-RPC protocol doesn't support returning None.

JordiBForgeFlow committed:
Corrección bug 1021631 por el cual la aplicación da el error "NameError: global name 'pooler' is not defined" al iniciar el calculo.

albertca committed:
l10n_es_aeat_mod349: Permitir seleccionar varios períodos en lugar de sólo uno y arrelgar un bug en la generación del modelo cuando el partner tiene el campo title lleno.

Commit:
[ADD] l10n_es_aeat_mod349: Añadimos dos nuevas claves de operación M y H que son respectivamente Entregas intracomunitarias con exención de tasas y Entregas intracomunitarias con exención de tasas por un representante fiscal oficial fueron publicadas el 26/11/2011 en el boe. Se puede ver el artículo en: http://planartus.blogspot.es/1309888740/

Commit:
Commit inicial de una rama para la versión 7.0. Todos los módulos fueron marcados como no instalables mientras no se compruebe que funcionan o migren, para poder usarlos o probarlos se debería poner el atributo installable de los ficheros __openerp__.py de cada módulo a True

acysos committed:
[FIX] l10n_es_aeat_mod349: string sin cerrar en l10n_es_aeat_mod349-pot

Commit:
Launchpad automatic translations update.

Commit:
[REF] cambiados los osv por orm.Model y orm.TransientModel, añadido en os import de objetos de OpenERP, el namespace openerp, refactorizado a PEP-8 y adaptacion vistas a formato 7

Commit:
Launchpad automatic translations update.

Commit:
[IMP] Adaptación de los import a formato v7/v8

Commit:
[IMP] l10n_es_aeat_mod349: Diversas mejoras y adaptación al nuevo l10n_es_aeat.

* Traducciones.
* Vista del wizard BOE
* Vistas de facturas mejoradas
* Icono
* Asignar una clave a las facturas ya existentes.

Commit:
[FIX] Adaptación a v8 de account_balance_reporting, l10n_es_aeat_mod349 y l10n_es_payment_order

Commit:
Launchpad automatic translations update.

pedrobaeza committed:
[MIG] l10n_es_aeat_mod349: Con nueva API

Commit:
[IMP] unificar campos en el módulo base + migrar a la nueva api.

omar7r committed:
[FIX] l10n_es_aeat_mod349: Se corrige un problema al calcular las facturas rectificativas a declarar.

pablocm-aserti committed:
[FIX] l10n_es_aeat_mod349: Uso incorrecto de ondelete en definición de o2m
[FIX] l10n_es_aeat_mod349: Añadida restricción que impide eliminar facturas que aparezcan en registros de detalle

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_mod349: 3 things:

* IMP: Fiscal position onchange in new API + speed imp + refactoring
* FIX: l10n_es_aeat_mod349: Remove check of names

  There's no need of a check about the name validity, because l10n_es_aeat
  already transform all the exported strings to valid ASCII characters
  (though unidecode), so I remove it entirely this check.

* FIX: Use first period year in export

  Relying in fiscal year code for getting the fiscal year is not safe,
  because you can use that code for defining your journal sequences (for
  example, with the module account_auto_fy_sequence).

  We can't use also the year of the fiscal year, because you can have
  split fiscal years (i. e. like the academic courses). What we uses then
  is the year of the first period of the declaration, as there's no
  legal possibility of having mixed periods from different years.

oca-transbot committed:
OCA Transbot updated translations from Transifex

antespi committed:
[FIX] l10n_es_aeat_mod349: Name field in 349 conflict + add _rec_name

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Don't depend on fiscal year code for the export

Roodin committed:
[FIX]. l10n_es_aeat_mod349 . Recuperar funcionalidad perdida para poder calcular el 349 en períodos distintos de los fiscales

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Search only one refund detail record

angelmoya committed:
[MIG] l10n_es_aeat_mod349

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_mod349

* Pylint
* Use new API

[skip ci]

luismontalba committed:
[IMP] l10n_es_aeat_mod349: Based on account moves

JordiBForgeFlow committed:
refactoring de l10n_es_aeat_mod349. Mejoras incluídas:
* Mejora del rendimiento.
* Apuntes contables como fuente de datos. El motor del informe se construye a partir de apuntes
  contables, y la información de facturas se consulta cuando se necesita. De este modo aseguramos que
  los resultados serán consistentes con los otros modelos de AEAT, que también toman como fuente de
  datos los apuntes contables. Así evitamos posibles inconsistencias.
* Nuevo informe Qweb, que sustituye al antiguo RML
* Correcciones para exporación a formato BOE. Permite importar el modelo a la web de hacienda.
* Permite borrar operaciones que no se quiere declarar. El usuario puede eliminar las operaciones que
  no desee incluir en la presentación. Esto será especialmente útil en el caso de presentaciones
  complementarias.
* En declaraciones normales y complementarias se excluyen automáticamente operaciones ya
  presentadas anteriormente.
* Script de migración para tener en cuenta los cambios en el modelo de datos.

JordiBForgeFlow committed:
[MIG] l10n_es_aeat_mod349: Migration to 10.0

pedrobaeza committed:
[FIX+IMP] l10n_es_aeat_mod349: Several fixes + imps to new approach

* Save in account.move.line directly the 349 export key via a selection field, so you are not bound to a map line record instead of mapping changes.
* Field renamings for better understanding.
* Map data with command `4` instead of `6` for allowing custom mappings.
* Added constraint for avoiding duplicated mapping lines with the same operation key.
* Fix tests asserts and code
* Fix refund records computation
* Complete behavior for imported move lines declarations
* Switch certain fields to computed ones, for avoiding extra code that can be handled easily with ORM, but without performance penalty thanks to norecompute ORM option.
* Remove useless check for invalid characters, as export now removes them

pedrobaeza committed:
[MIG] l10n_es_aeat_mod349: Migration to 11.0

Commit:
[FIX] l10n_es_aeat_mod349: Don't repeat refund details

carlosdauden committed:
[FIX] l10n_es_aeat_mod349: 2 things:

* Raise error with amount zero records (invoices+refunds=0.0)
* l10n_es_aeat_mod349: Missed by name param assignation precision_rounding

oca-travis committed:
[UPD] Update l10n_es_aeat_mod349.pot

JordiBForgeFlow committed:
l10n_es_aeat_349: corrige error en el 349 cuando se generan apuntes contables
asociados a una factura que no apuntan directamente a una línea de factura.

aitorbouzas committed:
[FIX] l10n_es_aeat_mod349: Fix report
Fix readme

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Adjust views permissions + remove old authors

* If views are not restricted to invoicing users, salesman won't be able to see the invoices.
* Current code doesn't share anything with original one, in approach nor lines of code, so
  original authors shouldn't be in the manifest key. I keep the copyright any way for
  historical reasons.

pedrobaeza committed:
[MIG] l10n_es_aeat_mod349: Migration to 12.0

* Standard procedure
* README by fragments

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Fix tax reference

Copy/paste operation left out one needed tax.

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Proper string format

pedrobaeza committed:
[IMP] l10n_es_aeat: Add active field to export configs

This way, we don't initially see all the subconfigs for each model and we can concentrate
on main export configurations. This is got through `active` field default behaviour.
Archived config are still directly searchable through the subconfig selector in the
configuration line.

All models export configs have been adapted to this new approach.

Commit:
[12.0][FIX] l10n_es_aeat_mod349: report

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Handle correctly Greek partners

As the VAT number should start with 'EL', not the country code 'GR'.

oihane committed:
[FIX] l10n_es_aeat_mod349: take all lines related to refunded invoice

josepmy committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (146 of 146 strings)

Translation: l10n-spain-12.0/l10n-spain-12.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-l10n_es_aeat_mod349/es/

Commit:
[FIX] l10n_es_aeat_mod349: refund groups

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Proper year and month grouping key when no refund link

Code was not adapted to new v12 date fields with Python objects.

Courtesy of Jesús Ramoneda

pedrobaeza committed:
[IMP] l10n_es_aeat: Add chatter + activities to AEAT models

* Include chatter (mail.thread)
* Include activities (mail.activity.mixin)
* Track AEAT report state change
* Include new admin user in AEAT group by default

oihane committed:
[FIX] l10n_es_aeat_mod349: month from python only with one digit period_type needs 2

jesusVMayor committed:
[FIX] l10n_es_aeat_mod349: Corregido cálculo de periodo en base al mes del apunte.

pedrobaeza committed:
[IMP] l10n_es_aeat_mod349: Remove children records when recovering report to draft

This way, we allow things like cancelling invoices that are included in the report.

pedrobaeza committed:
[IMP] l10n_es_aeat_mod349: Don't depends on tax description

Continuation of 611c24bc and later commits. This way, we can decouple tax label from
the matching, thanks to the XML-ID that Odoo gives to created taxes. We also
simplifies the code thanks to this.

josepmy committed:
Translated using Weblate (Spanish)

Currently translated at 82.5% (146 of 177 strings)

Translation: l10n-spain-12.0/l10n-spain-12.0-l10n_es_aeat_mod349
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-12-0/l10n-spain-12-0-l10n_es_aeat_mod349/es/

pedrobaeza committed:
[FIX] l10n_es_aeat_mod349: Refund record totals + details

**Steps to reproduce**

- Create an intra-community invoice with several lines in May.
- Create a 349 AEAT report for May and calculate it.
- Refund the invoice in June.
- Create a 349 report for June and calculate it.

**Current behavior**

You get double origin amount in the refund record.

**Expected behavior**

Get proper origin amount.

**Solution**

We have to pre-group refund details by origin invoice, and then process them in
batch for getting the total amount.

It includes regression test.

Commit:
[FIX] l10n_es_aeat_mod349: total refunds calculation

pedrobaeza committed:
[IMP] l10n_es_aeat_mod349: black, isort, prettier

pedrobaeza committed:
[MIG] l10n_es_aeat_mod349: Migration to 13.0

- General procedure
- In report change type > statement_type + migration script
- Computation adapted
- Adapt account.invoice > account.move in logic + migration script
- Adapt tests

oca-travis committed:
[UPD] Update l10n_es_aeat_mod349.pot

Jortolsa-S73 reviewed (approved):
Revisión funcional, comprobado en dos empresas presentando la declaración.

albariera reviewed (approved):
Technical review LGTM :+1:

etobella reviewed (approved):
Technical review.

HaraldPanten reviewed (approved):
Revisión funcional 👍

HaraldPanten commented:
@pedrobaeza si los ves OK, fusionamos

pedrobaeza commented:
Varios comentarios:

- No veo correcta la introducción de una nueva dependencia `account_invoice_refund_link`, y todo ese código de FactorLibre que aún no está verificado. No es buena idea depender de un enlace que no tenemos claro que vaya a estar rellenado (porque depende del momento de instalación del otro módulo).
- Hay un commit extra al final que debería hacerse squash.

Pediría que se elimine ese añadido, y se haga la migración de lo que hay en v13 ahora.

HaraldPanten commented:
@miguels73 Puedes atender los comentarios que comenta Pedro?

Gracias y saludos.

miguel-S73 commented:
@HaraldPanten Sí, voy a preparar el PR con la migración desde la 13.0. 
@pedrobaeza Había hecho un commit a parte con ese último cambio por si se necesita hacer un `cherry-pick` para otras versiones.

miguel-S73 committed:
[IMP] l10n_es_aeat_mod349: black, isort, prettier

HaraldPanten commented:
@miguels73 ¿Quieres que lo retomemos nosotros?

miguel-S73 commented:
@HaraldPanten , ya tengo los cambios en local me falta realizar una última prueba funcional y pasar los tests

HaraldPanten commented:
Perfecto!

miguel-S73 committed:
[MIG] l10n_es_aeat_mod349: migration to 14.0

miguel-S73 committed:
[FIX] l10n_es_aeat_mod349: origin amount incorrect

The calculation that was being done until now in the field origin_amount is incorrect since the total of the lines of the invoice to rectify was being added. The correct way is to add all the lines of the invoices of the period where the lines of the invoice to rectify are. Also, we must keep in mind that if there are intermediate periods between the original period and the period where the rectification is taking place, it's necessary to check if there is any rectification of the original period in between these periods. This happens in this way because the right original_amount  will be the value of the total_operation_amount corresponding to the last period found in between the periods

This commit has been based from this OCA/l10n-spain#1579

miguel-S73 commented:
@HaraldPanten changes applied

HaraldPanten commented:
@pedrobaeza por mí fusionamos. Lo que no tengo claro es si el FIX era mejor aquí o en otro PR. Pero en cualquier caso está separado en un commit aparte.

pedrobaeza commented:
/ocabot merge nobump
